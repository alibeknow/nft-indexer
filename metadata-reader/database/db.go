package database

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
	"go.mongodb.org/mongo-driver/mongo/writeconcern"
	"go.uber.org/zap"
)

type Config struct {
	Name             string
	Protocol         string
	Username         string
	Password         string
	Host             string
	Port             int
	SSLEnable        bool
	DirectConnection bool
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	if config.DirectConnection {
		url = fmt.Sprintf(
			"%s/?connect=direct",
			url,
		)
	}
	return url
}

type DB struct {
	log    *zap.Logger
	client *mongo.Client
	name   string
}

func Open(log *zap.Logger, ctx context.Context, config Config) (*DB, error) {
	opts := options.Client().ApplyURI(config.URL())

	if config.SSLEnable {
		path, _ := os.Getwd()
		pemPath := fmt.Sprintf("%s/../rds-combined-ca-bundle.pem", path)
		rootPEM, err := os.ReadFile(pemPath)
		if err != nil {
			return nil, err
		}

		roots := x509.NewCertPool()
		if ok := roots.AppendCertsFromPEM(rootPEM); !ok {
			return nil, fmt.Errorf("get certs from %s fail!\n", pemPath)
		}
		tlsConfig := &tls.Config{
			RootCAs:            roots,
			InsecureSkipVerify: true,
		}

		opts.SetReadPreference(readpref.Secondary())
		opts.SetWriteConcern(writeconcern.New(writeconcern.WMajority(), writeconcern.J(true), writeconcern.WTimeout(1000)))
		opts.SetTLSConfig(tlsConfig)
	}

	client, err := mongo.NewClient(opts)
	if err != nil {
		return nil, err
	}

	err = client.Connect(ctx)
	if err != nil {
		return nil, err
	}

	pingCtx, pingCancel := context.WithTimeout(ctx, 5*time.Second)
	defer pingCancel()
	err = client.Ping(pingCtx, nil)
	if err != nil {
		return nil, err
	}
	log.Info("Connected to DB", zap.String("url", config.URL()))

	return &DB{
		log:    log,
		client: client,
		name:   config.Name,
	}, nil
}

func (db *DB) Close() error {
	return db.client.Disconnect(context.Background())
}

func (db *DB) Metadata() *MetadataDB {
	coll := db.client.Database(db.name).Collection("metadata")
	return &MetadataDB{
		collection: coll,
	}
}
