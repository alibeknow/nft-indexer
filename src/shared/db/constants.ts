export const MONGO_TOKENS_COLLECTION_NAME = 'tokens';
export const MONGO_CONTRACTS_COLLECTION_NAME = 'contracts';
export const MONGO_METADATA_COLLECTION_NAME = 'metadata';
export const MONGO_EVENTS_COLLECTION_NAME = 'events';
export const MONGO_EVENTS_READER_STATE_LOG_COLLECTION_NAME = 'events_reader_state_log';
export const MONGO_BLOCKS_QUEUE_COLLECTION_NAME = 'blocks_queue';
export const MONGO_LAZY_MINT_BLOCKS_QUEUE_COLLECTION_NAME = 'lazy_mint_blocks_queue';
export const MONGO_SERVER_DUPLICATE_KEY_ERROR_CODE = 11000;
