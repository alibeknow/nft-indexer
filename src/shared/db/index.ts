export * from './constants';
export * from './db.class';
export * from './db.deprecated';
export * from './db.config';
export * from './db.module';
export * from './db.provider';
export * from './mongo-sort';

