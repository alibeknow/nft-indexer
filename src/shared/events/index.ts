export * from './events-repo';
export * from './state-log-repo';
