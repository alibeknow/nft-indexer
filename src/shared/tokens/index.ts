export * from './repo';
export * from './to-valid-hex';
export * from './token-metadata-status';
export * from './token-standard';
