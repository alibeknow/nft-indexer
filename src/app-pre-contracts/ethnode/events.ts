import { BigNumber, ethers } from 'ethers';
import * as retry from 'retry';
import { CONTRACT_OBJECTS } from '@shared/pre-721';
import { logger } from '@shared/logger';
import { TokenStandard } from '@shared/tokens';
import {
  ETH_NODE_GET_LOGS_MAX_TIMEOUT,
  ETH_NODE_GET_LOGS_MIN_TIMEOUT,
  ETH_NODE_GET_LOGS_RETRIES,
  ZERO_FROM_ADDRESS,
} from './constants';

export type TransferEvent = {
  blockNumber: number;
  logIndex: number;
  transactionHash: string;
  contractAddress: string;
  type: TokenStandard;
  from: string;
  to: string;
  ids: string[];
  values: number[];
};

type ContractInterface = {
  [key: string]: ethers.utils.Interface;
};

/**
* Asynchronously get logs matching the events filter
* @param {ethers.providers.JsonRpcProvider} provider custom provider
* @param {ethers.utils.Interface} iface interface instance from contract abi
* @param {number} blockNumberFrom start from block number
* @param {number} blockNumberTo end block number
*
* @returns {Promise<ethers.providers.Log[]>} Array of logs
*/
export async function getLogs(
  provider: ethers.providers.JsonRpcProvider, blockNumberFrom: number, blockNumberTo: number,
): Promise<ethers.providers.Log[]> {

  const address = [ ...CONTRACT_OBJECTS.map((contract) => {
    return contract.address;
  }) ];

  const filter = {
    fromBlock: fromDecToHex(blockNumberFrom),
    toBlock: fromDecToHex(blockNumberTo),
    address,
    topics: [
      [
        ethers.utils.id('Transfer(address,address,uint256)'),
        ethers.utils.id('TransferSingle(address,address,address,uint256,uint256)'),
        ethers.utils.id('TransferBatch(address,address,address,uint256[],uint256[])'),
      ],
    ],
  };

  const operation = retry.operation({
    retries: ETH_NODE_GET_LOGS_RETRIES,
    minTimeout: ETH_NODE_GET_LOGS_MIN_TIMEOUT,
    maxTimeout: ETH_NODE_GET_LOGS_MAX_TIMEOUT,
  });

  return new Promise((resolve, reject) => {
    operation.attempt(async (currentAttempt: number) => {
      try {
        logger.info({ from: blockNumberFrom, to: blockNumberTo, msg: `Attempt #${currentAttempt}: Requesting ETH node logs...` });
        const rawLogs = await provider.send('eth_getLogs', [ filter ]);

        const logs: ethers.providers.Log[] = [];

        for (const rawLog of rawLogs) {
          const {
            blockNumber,
            blockHash,
            transactionIndex,
            removed,
            address,
            data,
            topics,
            transactionHash,
            logIndex,
          } = rawLog;

          const log: ethers.providers.Log = {
            blockNumber: fromHexToDec(blockNumber),
            transactionIndex: fromHexToDec(transactionIndex),
            address: ethers.utils.getAddress(address),
            logIndex: fromHexToDec(logIndex),
            blockHash,
            removed,
            data,
            topics,
            transactionHash,
          };

          logs.push(log);
        }

        resolve(logs);
      } catch (e) {
        if (operation.retry(e as unknown as Error)) {
          logger.error({
            from: blockNumberFrom,
            to: blockNumberTo,
            msg: `Error requesting ETH node logs (Attempt #${currentAttempt}): ${e}. Retrying...`,
          });

          return;
        }

        reject(operation.mainError());
      }
    });
  });
}

/**
* Asynchronously get transfer events that matches the log filter and push to array by events topics
* @param {number} blockNumberFrom start from block number
* @param {number} blockNumberTo end block number
* @param {ethers.providers.BaseProvider} provider custom provider
*
* @returns {Promise<TransferEvent[]>} Array of logs
*/
export async function getEthNodeTransferEvents(
  blockNumberFrom: number, blockNumberTo: number, provider: ethers.providers.Web3Provider,
): Promise<TransferEvent[]> {
  const pre721: ContractInterface = {};

  for (const contractObject of CONTRACT_OBJECTS) {
    pre721[contractObject.address] = new ethers.utils.Interface(contractObject.abi);
  }

  const logs = await getLogs(provider, blockNumberFrom, blockNumberTo);

  const res: TransferEvent[] = [];
  for (const log of logs) {
    const iface = pre721[log.address];

    const parsedLog = iface.parseLog(log);
    if (parsedLog.args.from !== ZERO_FROM_ADDRESS) {
      continue;
    }

    const ids: string[] = [];
    const values: number[] = [];
    let type: TokenStandard;

    switch (parsedLog.name) {
    case 'Transfer':
      ids.push(parseTokenId(parsedLog.args.tokenId));
      values.push(1);
      type = TokenStandard.PRE721;
      break;
    case 'TransferSingle':
      ids.push(parseTokenId(parsedLog.args.id));
      values.push(parseTokenValue(parsedLog.args.value));
      type = TokenStandard.PRE1151;
      break;
    case 'TransferBatch':
      if (parsedLog.args && parsedLog.args.length) {
        ids.push(...parsedLog.args[parsedLog.args.length - 2].map(parseTokenId));
        values.push(...parsedLog.args[parsedLog.args.length - 1].map(parseTokenValue));
        // values keyword is overlapping with ethers function values, so it's the only way to get proper values
      }
      type = TokenStandard.PRE1151;
      break;
    default:
      throw new Error('Not suitable log. Log name is not Transfer, TransferSingle or TransferBatch');
    }

    res.push({
      blockNumber: log.blockNumber,
      logIndex: log.logIndex,
      transactionHash: log.transactionHash,
      contractAddress: log.address,
      from: parsedLog.args.from,
      to: parsedLog.args.to,
      type,
      ids,
      values,
    });
  }

  return res;
}

/**
* Convert id to hex string
* @param {BigNumber} id
*
* @returns {string} hex string
*/
function parseTokenId(id: BigNumber): string {
  return id.toHexString();
}

/**
* Convert token value to number
* @param {BigNumber} value
*
* @returns {number} value to number
*/
function parseTokenValue(value: BigNumber): number {
  return +value.toString();
}

function fromHexToDec(value: string): number {
  return BigNumber.from(value).toNumber();
}

function fromDecToHex(value: number): string {
  return ethers.utils.hexValue(value);
}
