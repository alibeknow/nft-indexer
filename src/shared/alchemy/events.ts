import { BigNumber, ethers } from 'ethers';
import * as retry from 'retry';
import { logger } from '@shared/logger';
import { ZERO_FROM_ADDRESS } from '@shared/nfts';
import { TokenStandard } from '@shared/tokens';
import {
  ALCHEMY_GET_ASSET_TRANSFERS_JRPC_METHOD,
  ALCHEMY_GET_EVENTS_MAX_TIMEOUT,
  ALCHEMY_GET_EVENTS_MIN_TIMEOUT,
  ALCHEMY_GET_EVENTS_RETRIES,
} from './constants';

type RawContract = {
  address: string;
};

type ERC1115Metadata = {
  tokenId: string;
  value: BigNumber;
};

type Transfer = {
  blockNum: number;
  hash: string;
  from: string;
  to: string;
  erc721TokenId: string;
  erc1155Metadata: ERC1115Metadata[];
  rawContract: RawContract;
};

type AlchemyTransferEventResults = {
  transfers: Transfer[];
  pageKey: string;
};

type RequestParams = {
  fromBlock: string;
  toBlock: string;
  category: Array<string>;
  fromAddress: string;
  pageKey?: string;
};

export type EventResponse = {
  blockNumber: number;
  contractAddress: string;
  ids: string[];
  values: number[];
  type: TokenStandard;
};

/**
* Asynchronously request transfer events from Alchemy
* @param {provider} provider json rpc provider
* @param {number} blockNumberFrom start from block number
* @param {number} blockNumberTo end block number
* @param {string} pageKey
*
* @returns {Promise<AlchemyTransferEventResults>} transfer event data
*/
export async function requestTransferEvents(
  provider: ethers.providers.JsonRpcProvider,
  blockNumberFrom: number,
  blockNumberTo: number,
  pageKey?: string,
): Promise<AlchemyTransferEventResults> {
  let requestParams: RequestParams = {
    fromBlock: BigNumber.from(blockNumberFrom).toHexString(),
    toBlock: BigNumber.from(blockNumberTo).toHexString(),
    fromAddress: ZERO_FROM_ADDRESS,
    category: [
      'erc721',
      'erc1155',
    ],
  };

  if(pageKey) {
    requestParams = { ...requestParams, pageKey };
  }

  const operation = retry.operation({
    retries: ALCHEMY_GET_EVENTS_RETRIES,
    minTimeout: ALCHEMY_GET_EVENTS_MIN_TIMEOUT,
    maxTimeout: ALCHEMY_GET_EVENTS_MAX_TIMEOUT,
  });

  return new Promise((resolve, reject) => {
    operation.attempt(async (currentAttempt: number) => {
      try {
        logger.info({ from: blockNumberFrom, to: blockNumberTo, msg: `Attempt #${currentAttempt}: requesting Alchemy events with params ${JSON.stringify(requestParams)}` });
        const results = await provider.send(ALCHEMY_GET_ASSET_TRANSFERS_JRPC_METHOD, [ requestParams ]);
        resolve(results);
      } catch (e) {
        if(operation.retry(e as unknown as Error)) {
          logger.error({
            msg: `Error requesting Alchemy events (Attempt #${currentAttempt}). Retrying...`,
            from: blockNumberFrom,
            to: blockNumberTo,
            requestParams: JSON.stringify(requestParams),
            error: e,
          });

          return;
        }

        reject(operation.mainError());
      }
    });
  });
}

/**
* Asynchronously get transfer events data from Alchemy
* @param {number} blockNumberFrom start from block number
* @param {number} blockNumberTo end block number
* @param {provider} provider json rpc provider
*
* @returns {Promise<EventResponse[]>} Transfer event data response
*/
export async function getAlchemyTransferEvents(blockNumberFrom: number, blockNumberTo: number, provider: ethers.providers.JsonRpcProvider): Promise<EventResponse[]> {
  let transfers: Transfer[] = [];
  let results = await requestTransferEvents(provider, blockNumberFrom, blockNumberTo);
  transfers = transfers.concat(results.transfers);

  while (results.pageKey) {
    results = await requestTransferEvents(provider, blockNumberFrom, blockNumberTo, results.pageKey);
    transfers = transfers.concat(results.transfers);
  }

  return transfers.map((item) => {
    const ids: string[] = [];
    const values: number[] = [];
    let type: TokenStandard;

    if (item.erc721TokenId) {
      ids.push(item.erc721TokenId);
      values.push(1);
      type = TokenStandard.ERC721;
    } else {
      item.erc1155Metadata.forEach(m => {
        ids.push(m.tokenId);
        values.push(+m.value.toString());
      });
      type = TokenStandard.ERC1155;
    }

    return {
      blockNumber: item.blockNum,
      contractAddress: item.rawContract.address,
      ids,
      values,
      type,
    };
  });
}
