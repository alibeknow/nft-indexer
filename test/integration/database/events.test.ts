import { DBClass, IDatabaseConfig, getDatabaseConfig } from '@shared/db';
import { EventsRepo } from '@shared/events';
import { TokenStandard } from '@shared/tokens';

describe('Events repo', () => {
  let db: DBClass;
  beforeAll(async () => {
    const dbConfig: IDatabaseConfig = getDatabaseConfig();
    db = new DBClass({ ...dbConfig, db: `db-events-repo-test-${dbConfig.db}` });
    await db.open();
  });

  describe('[bulkSave] method', () => {
    it('should save batch of events in events collection', async () => {
      const eventsRepo = new EventsRepo(db);
      const saveEventsResult = await eventsRepo.bulkSave([ {
        blockNumber: 14306496,
        logIndex: 193,
        transactionHash: '0xfd1ac0250ccc5719cf31e55865141c1c1a746a4933f04a2545fd70a0be984bb2',
        contractAddress: '0xc4545a0f49e1246bdba8601fd79412d240cfc6aa',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'fake-id-1', 'fake-id-2' ],
        values: [ 2, 4 ],
      }, {
        blockNumber: 14306497,
        logIndex: 1,
        transactionHash: '0x865141c1c1a746a4933f04a2545fd70a0be984bb2fd1ac0250ccc5719cf31e55',
        contractAddress: '0xfd79412d240cfc6aac4545a0f49e1246bdba8601',
        type: TokenStandard.ERC1155,
        from: '',
        to: '',
        ids: [ 'fake-id' ],
        values: [ 2 ],
      }, {
        blockNumber: 14306498,
        logIndex: 25,
        transactionHash: '0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2',
        contractAddress: '0x601fd79412dc4545aa8240cfc6aa0f49e1246bdb',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'some-id' ],
        values: [ 1 ],
      } ]);

      const docs = await db.events().find({ _id: { $in: [
        '14306496:0xfd1ac0250ccc5719cf31e55865141c1c1a746a4933f04a2545fd70a0be984bb2:193',
        '14306497:0x865141c1c1a746a4933f04a2545fd70a0be984bb2fd1ac0250ccc5719cf31e55:1',
        '14306498:0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2:25',
      ] } }).toArray();

      // To get rid of non-permanent field createdAt
      // eslint-disable-next-line
      const docsWithoutCreatedAt = docs.map(({ createdAt,  ...rest }) => ({ ...rest }));

      expect(docsWithoutCreatedAt).toMatchSnapshot();
      expect(saveEventsResult).toMatchSnapshot();
    });
  });

  describe('[getByBlockRange] method', () => {
    it('should return event documents by block range ordered by createdAt', async () => {
      const eventsRepo = new EventsRepo(db);
      await eventsRepo.bulkSave([ {
        blockNumber: 14306496,
        logIndex: 193,
        transactionHash: '0xfd1ac0250ccc5719cf31e55865141c1c1a746a4933f04a2545fd70a0be984bb2',
        contractAddress: '0xc4545a0f49e1246bdba8601fd79412d240cfc6aa',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'fake-id-1', 'fake-id-2' ],
        values: [ 2, 4 ],
      }, {
        blockNumber: 14306496,
        logIndex: 1,
        transactionHash: '0x865141c1c1a746a4933f04a2545fd70a0be984bb2fd1ac0250ccc5719cf31e55',
        contractAddress: '0xfd79412d240cfc6aac4545a0f49e1246bdba8601',
        type: TokenStandard.ERC1155,
        from: '',
        to: '',
        ids: [ 'fake-id' ],
        values: [ 2 ],
      }, {
        blockNumber: 14306498,
        logIndex: 25,
        transactionHash: '0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2',
        contractAddress: '0x601fd79412dc4545aa8240cfc6aa0f49e1246bdb',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'some-id' ],
        values: [ 1 ],
      } ]);

      const firstBlockDocs = await eventsRepo.getByBlockRange(14306496, 14306497);
      const secondBlockDocs = await eventsRepo.getByBlockRange(14306498, 14306499);

      // To get rid of non-permanent field createdAt
      // eslint-disable-next-line
      const firstBlockDocsWithoutCreatedAt = firstBlockDocs.map(({ createdAt,  ...rest }) => ({ ...rest }));
      // eslint-disable-next-line
      const secondBlockDocsWithoutCreatedAt = secondBlockDocs.map(({ createdAt,  ...rest }) => ({ ...rest }));

      expect(firstBlockDocs).toHaveLength(2);
      expect(firstBlockDocsWithoutCreatedAt).toMatchSnapshot();
      expect(secondBlockDocs).toHaveLength(1);
      expect(secondBlockDocsWithoutCreatedAt).toMatchSnapshot();
    });
  });

  describe('[deleteWhereBlockNumberGreaterThan] method', () => {
    it('should delete events where block greater than a provided block', async () => {
      const eventsRepo = new EventsRepo(db);
      await eventsRepo.bulkSave([ {
        blockNumber: 14306496,
        logIndex: 193,
        transactionHash: '0xfd1ac0250ccc5719cf31e55865141c1c1a746a4933f04a2545fd70a0be984bb2',
        contractAddress: '0xc4545a0f49e1246bdba8601fd79412d240cfc6aa',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'fake-id-1', 'fake-id-2' ],
        values: [ 2, 4 ],
      }, {
        blockNumber: 14306496,
        logIndex: 1,
        transactionHash: '0x865141c1c1a746a4933f04a2545fd70a0be984bb2fd1ac0250ccc5719cf31e55',
        contractAddress: '0xfd79412d240cfc6aac4545a0f49e1246bdba8601',
        type: TokenStandard.ERC1155,
        from: '',
        to: '',
        ids: [ 'fake-id' ],
        values: [ 2 ],
      }, {
        blockNumber: 14306498,
        logIndex: 25,
        transactionHash: '0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2',
        contractAddress: '0x601fd79412dc4545aa8240cfc6aa0f49e1246bdb',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'some-id' ],
        values: [ 1 ],
      }, {
        blockNumber: 14306499,
        logIndex: 25,
        transactionHash: '0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2',
        contractAddress: '0x601fd79412dc4545aa8240cfc6aa0f49e1246bdb',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'some-id' ],
        values: [ 1 ],
      } ]);

      const docsBeforeDeletion = await db.events().find().toArray();
      await eventsRepo.deleteWhereBlockNumberGreaterThan(14306498);
      const docs = await db.events().find().toArray();

      expect(docsBeforeDeletion).toHaveLength(4);
      expect(docs).toHaveLength(3);
    });
  });

  describe('[deleteByBlockRange] method', () => {
    it('should delete events within block range', async () => {
      const eventsRepo = new EventsRepo(db);
      await eventsRepo.bulkSave([ {
        blockNumber: 14306496,
        logIndex: 193,
        transactionHash: '0xfd1ac0250ccc5719cf31e55865141c1c1a746a4933f04a2545fd70a0be984bb2',
        contractAddress: '0xc4545a0f49e1246bdba8601fd79412d240cfc6aa',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'fake-id-1', 'fake-id-2' ],
        values: [ 2, 4 ],
      }, {
        blockNumber: 14306496,
        logIndex: 1,
        transactionHash: '0x865141c1c1a746a4933f04a2545fd70a0be984bb2fd1ac0250ccc5719cf31e55',
        contractAddress: '0xfd79412d240cfc6aac4545a0f49e1246bdba8601',
        type: TokenStandard.ERC1155,
        from: '',
        to: '',
        ids: [ 'fake-id' ],
        values: [ 2 ],
      }, {
        blockNumber: 14306498,
        logIndex: 25,
        transactionHash: '0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2',
        contractAddress: '0x601fd79412dc4545aa8240cfc6aa0f49e1246bdb',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'some-id' ],
        values: [ 1 ],
      }, {
        blockNumber: 14306499,
        logIndex: 25,
        transactionHash: '0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2',
        contractAddress: '0x601fd79412dc4545aa8240cfc6aa0f49e1246bdb',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'some-id' ],
        values: [ 1 ],
      } ]);

      const docsBeforeDeletion = await db.events().find().toArray();
      await eventsRepo.deleteByBlockRange(14306496, 14306498);
      const docs = await db.events().find().toArray();

      expect(docsBeforeDeletion).toHaveLength(4);
      expect(docs).toHaveLength(1);
    });
  });

  describe('[countByBlockRange] method', () => {
    it('should return number of events within block range', async () => {
      const eventsRepo = new EventsRepo(db);
      await eventsRepo.bulkSave([ {
        blockNumber: 14306496,
        logIndex: 193,
        transactionHash: '0xfd1ac0250ccc5719cf31e55865141c1c1a746a4933f04a2545fd70a0be984bb2',
        contractAddress: '0xc4545a0f49e1246bdba8601fd79412d240cfc6aa',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'fake-id-1', 'fake-id-2' ],
        values: [ 2, 4 ],
      }, {
        blockNumber: 14306496,
        logIndex: 1,
        transactionHash: '0x865141c1c1a746a4933f04a2545fd70a0be984bb2fd1ac0250ccc5719cf31e55',
        contractAddress: '0xfd79412d240cfc6aac4545a0f49e1246bdba8601',
        type: TokenStandard.ERC1155,
        from: '',
        to: '',
        ids: [ 'fake-id' ],
        values: [ 2 ],
      }, {
        blockNumber: 14306498,
        logIndex: 25,
        transactionHash: '0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2',
        contractAddress: '0x601fd79412dc4545aa8240cfc6aa0f49e1246bdb',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'some-id' ],
        values: [ 1 ],
      }, {
        blockNumber: 14306499,
        logIndex: 25,
        transactionHash: '0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2',
        contractAddress: '0x601fd79412dc4545aa8240cfc6aa0f49e1246bdb',
        type: TokenStandard.ERC721,
        from: '',
        to: '',
        ids: [ 'some-id' ],
        values: [ 1 ],
      } ]);

      const result1 = await eventsRepo.countByBlockRange(14306496, 14306498);
      const result2 = await eventsRepo.countByBlockRange(1, 15306498);
      const result3 = await eventsRepo.countByBlockRange(14306498, 14306498);
      const result4 = await eventsRepo.countByBlockRange(14306496, 14306496);

      expect(result1).toBe(3);
      expect(result2).toBe(4);
      expect(result3).toBe(1);
      expect(result4).toBe(2);
    });
  });

  afterEach(async () => {
    await db.events().deleteMany({});
  });

  afterAll(async () => {
    await db.close(true);
  });
});
