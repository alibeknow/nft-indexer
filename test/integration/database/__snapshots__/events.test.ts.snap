// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events repo [bulkSave] method should save batch of events in events collection 1`] = `
[
  {
    "_id": "14306496:0xfd1ac0250ccc5719cf31e55865141c1c1a746a4933f04a2545fd70a0be984bb2:193",
    "blockNumber": 14306496,
    "contractAddress": "0xc4545a0f49e1246bdba8601fd79412d240cfc6aa",
    "from": "",
    "ids": [
      "fake-id-1",
      "fake-id-2",
    ],
    "logIndex": 193,
    "to": "",
    "type": "ERC-721",
    "values": [
      2,
      4,
    ],
  },
  {
    "_id": "14306497:0x865141c1c1a746a4933f04a2545fd70a0be984bb2fd1ac0250ccc5719cf31e55:1",
    "blockNumber": 14306497,
    "contractAddress": "0xfd79412d240cfc6aac4545a0f49e1246bdba8601",
    "from": "",
    "ids": [
      "fake-id",
    ],
    "logIndex": 1,
    "to": "",
    "type": "ERC-1155",
    "values": [
      2,
    ],
  },
  {
    "_id": "14306498:0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2:25",
    "blockNumber": 14306498,
    "contractAddress": "0x601fd79412dc4545aa8240cfc6aa0f49e1246bdb",
    "from": "",
    "ids": [
      "some-id",
    ],
    "logIndex": 25,
    "to": "",
    "type": "ERC-721",
    "values": [
      1,
    ],
  },
]
`;

exports[`Events repo [bulkSave] method should save batch of events in events collection 2`] = `
{
  "insertedIds": [
    {
      "_id": "14306496:0xfd1ac0250ccc5719cf31e55865141c1c1a746a4933f04a2545fd70a0be984bb2:193",
      "index": 0,
    },
    {
      "_id": "14306497:0x865141c1c1a746a4933f04a2545fd70a0be984bb2fd1ac0250ccc5719cf31e55:1",
      "index": 1,
    },
    {
      "_id": "14306498:0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2:25",
      "index": 2,
    },
  ],
  "nInserted": 3,
  "nMatched": 0,
  "nModified": 0,
  "nRemoved": 0,
  "nUpserted": 0,
  "ok": 1,
  "upserted": [],
  "writeConcernErrors": [],
  "writeErrors": [],
}
`;

exports[`Events repo [getByBlockRange] method should return event documents by block range ordered by createdAt 1`] = `
[
  {
    "_id": "14306496:0x865141c1c1a746a4933f04a2545fd70a0be984bb2fd1ac0250ccc5719cf31e55:1",
    "blockNumber": 14306496,
    "contractAddress": "0xfd79412d240cfc6aac4545a0f49e1246bdba8601",
    "from": "",
    "ids": [
      "fake-id",
    ],
    "logIndex": 1,
    "to": "",
    "type": "ERC-1155",
    "values": [
      2,
    ],
  },
  {
    "_id": "14306496:0xfd1ac0250ccc5719cf31e55865141c1c1a746a4933f04a2545fd70a0be984bb2:193",
    "blockNumber": 14306496,
    "contractAddress": "0xc4545a0f49e1246bdba8601fd79412d240cfc6aa",
    "from": "",
    "ids": [
      "fake-id-1",
      "fake-id-2",
    ],
    "logIndex": 193,
    "to": "",
    "type": "ERC-721",
    "values": [
      2,
      4,
    ],
  },
]
`;

exports[`Events repo [getByBlockRange] method should return event documents by block range ordered by createdAt 2`] = `
[
  {
    "_id": "14306498:0x31e55865141c1c1a746a4fd1ac0250ccc5719cf933f04a2545fd70a0be984bb2:25",
    "blockNumber": 14306498,
    "contractAddress": "0x601fd79412dc4545aa8240cfc6aa0f49e1246bdb",
    "from": "",
    "ids": [
      "some-id",
    ],
    "logIndex": 25,
    "to": "",
    "type": "ERC-721",
    "values": [
      1,
    ],
  },
]
`;
